File: GeeksForGeeks/DSA_Series/Day4-BinarySearch/p2.cpp
#include<iostream>
#include<bits/stdc++.h>

using namespace std;
class Solution:
    public:
    int binarySearch(int arr[], int low, int high, int x){
        while (low <= high){
            int mid = (low + high) / 2;
             
            if (arr[mid] == x)
                return mid;
            else if (arr[mid] > x)
                high = mid - 1;
            else   
                low = mid + 1;   
        }
    };
    string isKSortedArray(int arr[], int n, int k)
    {
        //code here.
        int temp[n];
        for(int i=0; i<n;i++){
            temp[i]=arr[i];
        }
        
        sort(temp,temp+n);
        int j;
        for (int i = 0; i<n; i++){
            j = binarySearch(temp, 0, n-1, arr[i]);
            // cout << j << endl;
            if(abs(i-j)>k){
                return "No";
            }
        }
        return "Yes";
    };

File: GeeksForGeeks/DSA_Series/Day4-BinarySearch/p2.py
## K sorted array

def isKSortedArray(arr, n, k): 
    #code here.
    def binary_search(arr, start, end, target):
        if end>=start:
            mid = start + (end-start)//2
            if (mid==0 or target>arr[mid-1]) and arr[mid]==target:
                return mid
            elif target>arr[mid]:
                return binary_search(arr,mid+1,end,target)
            else:
                return binary_search(arr, start, mid-1, target)
                
        return -1
        
    temp = sorted(arr)
    
    i=0
    while i<n:
        find_pos = binary_search(temp,0,n-1,arr[i])
        #print(find_pos,i,arr[i])
        if abs(find_pos-i)>k:
            return "No"
        i+=1
    
    return "Yes"

test_cases = [
    ([3, 2, 1, 5, 6, 4],6,2),
    ([3, 2, 1, 5, 6, 4],6,1),
    ([13, 8, 10, 7, 15, 14, 12],7,1),
    ([13, 8, 10, 7, 15, 14, 12],7,4)
    ]
for x,y,z in test_cases:
    print(isKSortedArray(x,y,z))